src/ncp/ncp_spinel.cpp

```diff
@@ -236,6 +394,181 @@ spinel_tid_t NcpSpinel::GetNextTid(void)
+}
+
+void NcpSpinel::GetFlagsFromSecurityPolicy(const otSecurityPolicy *aSecurityPolicy,
+                                           uint8_t                *aFlags,
+                                           uint8_t                 aFlagsLength)
+{
+    static constexpr uint8_t kObtainNetworkKeyMask        = 1 << 7; // Byte 0
+    static constexpr uint8_t kNativeCommissioningMask     = 1 << 6; // Byte 0
+    static constexpr uint8_t kRoutersMask                 = 1 << 5; // Byte 0
+    static constexpr uint8_t kExternalCommissioningMask   = 1 << 4; // Byte 0
+    static constexpr uint8_t kCommercialCommissioningMask = 1 << 2; // Byte 0
+    static constexpr uint8_t kAutonomousEnrollmentMask    = 1 << 1; // Byte 0
+    static constexpr uint8_t kNetworkKeyProvisioningMask  = 1 << 0; // Byte 0
+
+    static constexpr uint8_t kTobleLinkMask     = 1 << 7; // Byte 1
+    static constexpr uint8_t kNonCcmRoutersMask = 1 << 6; // Byte 1
+    static constexpr uint8_t kReservedMask      = 0x38;   // Byte 1
+
+    VerifyOrExit(aFlagsLength > 1);
+
+    memset(aFlags, 0, aFlagsLength);
+
+    if (aSecurityPolicy->mObtainNetworkKeyEnabled)
+    {
+        aFlags[0] |= kObtainNetworkKeyMask;
+    }
+
+    if (aSecurityPolicy->mNativeCommissioningEnabled)
+    {
+        aFlags[0] |= kNativeCommissioningMask;
+    }
+
+    if (aSecurityPolicy->mRoutersEnabled)
+    {
+        aFlags[0] |= kRoutersMask;
+    }
+
+    if (aSecurityPolicy->mExternalCommissioningEnabled)
+    {
+        aFlags[0] |= kExternalCommissioningMask;
+    }
+
+    if (!aSecurityPolicy->mCommercialCommissioningEnabled)
+    {
+        aFlags[0] |= kCommercialCommissioningMask;
+    }
+
+    if (!aSecurityPolicy->mAutonomousEnrollmentEnabled)
+    {
+        aFlags[0] |= kAutonomousEnrollmentMask;
+    }
+
+    if (!aSecurityPolicy->mNetworkKeyProvisioningEnabled)
+    {
+        aFlags[0] |= kNetworkKeyProvisioningMask;
+    }
+
+    VerifyOrExit(aFlagsLength > sizeof(aFlags[0]));
+
+    if (aSecurityPolicy->mTobleLinkEnabled)
+    {
+        aFlags[1] |= kTobleLinkMask;
+    }
+
+    if (!aSecurityPolicy->mNonCcmRoutersEnabled)
+    {
+        aFlags[1] |= kNonCcmRoutersMask;
+    }
+
+    aFlags[1] |= kReservedMask;
+    aFlags[1] |= aSecurityPolicy->mVersionThresholdForRouting;
+
+exit:
+    return;
+}
+
+otError NcpSpinel::EncodeDatasetSetActiveTlvs(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs)
+{
+    otError              error = OT_ERROR_NONE;
+    otOperationalDataset dataset;
+    otIp6Address         addrMlPrefix;
+    uint8_t              flagsSecurityPolicy[2];
+
+    SuccessOrExit(error = otDatasetParseTlvs(&aActiveOpDatasetTlvs, &dataset));
+    GetFlagsFromSecurityPolicy(&dataset.mSecurityPolicy, flagsSecurityPolicy, sizeof(flagsSecurityPolicy));
+    memcpy(addrMlPrefix.mFields.m8, dataset.mMeshLocalPrefix.m8, sizeof(otMeshLocalPrefix));
+    memset(addrMlPrefix.mFields.m8 + 8, 0, sizeof(otMeshLocalPrefix));
+
+    SuccessOrExit(error = mEncoder.OpenStruct());
```

Comment:
Ideally, we should always transport Operational Datasets as TLVs. In particular, it should be possible to configure an Operational Dataset where the TLVs are not defined in the current Thread Specification.

Reply:
Make sense! I've checked the existing spinel properties and currently there is no TLV version of `SPINEL_PROP_THREAD_ACTIVE_DATASET` or `SPINEL_PROP_THREAD_PENDING_DATASET`. So we first need to add new spinel properties  `SPINEL_PROP_THREAD_ACTIVE_DATASET_TLV` and the handling on NCP side. Shall we use another PR to do the change after I add the new properties? I'll create an issue to track this.